Add missing size prefix for serialize_bytes

okready opened the pull request:
Serializer::serialize_bytes does not write the buffer size when
serializing, preventing deserialization from working properly when
serializing types that make use of this method (e.g. Bytes and BytesMut
from the bytes crate, CStr and CString from std::ffi). Simply adding the
missing size serialization breaks all size prefix serialization, as size
prefix serialization uses the existing Serializer::serialize_bytes
implementation to write the VarintUsize bytes to the output flavor.

This commit adds a try_push_varint_usize method to the SerFlavor trait
for writing out the encoded VarintUsize bytes, replacing all
VarintUsize::serialize calls with the SerFlavor method call, and adds
the same call to Serializer::serialize_bytes to ensure the size prefix
is now properly written. The VarintUsize::serialize implementation has
also been updated to serialize the encoded bytes as a tuple, mostly
retaining compatibility with the previous implementation while using
postcard's serializer, but potentially breaking compatibility with other
serializers. A deserialization test has also been added for
serialize_bytes output to mitigate possible regressions.

Fixes #21.

jamesmunns commented:
Hi @okready! Thanks for taking care of this. To be clear - this doesn't change the "on the wire" representation for any values, correct? I admit I don't totally grok what the problem was/what this fixes, but that's probably my lack of understanding of this part of the serde machinery!

okready commented:
Yeah, I re-read my explanation and realized it's pretty bad. :) In short, the wire representation should remain the same.

To try to clarify, `Serializer::serialize_bytes` essentially allows for serializing slices of bytes without making separate individual calls to `serialize_seq` and `serialize_element`, so it needs to include the length when serializing, otherwise we break serialization of any types that use it (such as the `bytes` crate and `CStr`/`CString`). Unfortunately, you can't just serialize a `VarintUsize` at the top of the `serialize_bytes` call since `VarintUsize::serialize` itself calls `serialize_bytes`, so you also need to change how varints are written.

Simply changing the `Serialize` implementation for `VarintUsize` is pretty limiting since you're abstracted from the `Serializer` internals, so instead I added a method to `SerFlavor` to directly write the varint bytes to the output buffer. It has a default implementation, so existing code should still compile against it the same, but I believe it still warrants a minor version bump according to semver rules.

The "potentially breaking compatibility with other serializers" statement in my original commit message was because I mistakenly thought `VarintUsize` was public at the time (I saw the `pub` scope for the type but didn't look at the scope of the module itself). Since it isn't public, I've removed its `Serialize` implementation for it since it's no longer used, but I can add one back if you'd prefer. I've also updated my original commit message to try to make things clearer.

okready commented:
It should be possible to add a regression test that verifies the wire format for various types hasn't changed. I wouldn't be able to get to it today, put should have time tomorrow or later on this week.

okready commented:
I realized the `loopback` test would have failed if the wire format changed, so it looks like that should cover backwards compatibility.

I've added in a new commit though to test against `CString` within the `loopback` test to verify the `serialize_bytes`/`deserialize_byte_buf` implementations work. It's limited to the `use-std` feature and required flipping on the `serde/std` feature under that configuration (AFAIK you can't specify test-only features, but I can remove that test or put in something different if there's a reason not to enable `serde/std` when `use-std` is enabled).

jamesmunns commented:
Thanks @okready! I'll release this shortly.
